WordPress/gutenberg
# Query Loop: Add animation when navigating with "enhanced pagination" enabled

## What?

This PR adds a subtle animation when visiting different pages of a Query Loop with "enhanced pagination" enabled. It also adds a smooth scroll behavior for the navigation.

When `reduced-motion` is preferred, the animations and the smooth scroll are turned off automatically.

Will close https://github.com/WordPress/gutenberg/issues/54060.

## Why?

To make the transition a bit more _kansei_, as mentioned in https://github.com/WordPress/gutenberg/pull/53812#issuecomment-1693305042.

## How?

For the animations, I had to add a new option to `navigate()` named `beforeRender`. It's a callback that runs right before we call Preact's render to update the HTML. The callback can be asynchronous, and this PR uses it to animate the current Post Template block before it's updated.

For the scroll behavior, I'm changing the value of the `scroll-behavior` CSS  property on the [`document.scrollingElement`](https://developer.mozilla.org/en-US/docs/Web/API/document/scrollingElement) to `smooth` right before calling `element.focus()` and restoring its previous value right after.

## Testing Instructions

1. Go to the site editor.
2. In the Blog Home template, enable the "enhanced pagination" setting for the Query Loop block.
3. Ensure you have a Query Pagination block.
4. Visit the home page.
5. Click on Older Posts.
6. Check that the fade-out / fade-in animations occur.
7. Check that the browser smoothly scrolls to the focused item (the first link).
8. Go to your OS accessibility settings and enable "reduce motion".
5. Click on Older Posts.
6. Check that the fade-out / fade-in animations don't happen this time.
7. Check that the scroll jumps directly to the focused item (the first link).

## Screenshots or screencast


| Without `reduced-motion` | With `reduced-motion` |
| --- | --- |
| <video src="https://github.com/WordPress/gutenberg/assets/6917969/f51c151c-16aa-4f63-8eca-a684009acae1"></video> | <video src="https://github.com/WordPress/gutenberg/assets/6917969/285b28a8-84fd-46bb-8b72-1e6a27066b4a" ></video> |
